src/goto-symex/goto_symex.cpp

```diff
@@ -665,3 +681,179 @@ bool goto_symext::constant_propagate_delete_char_at(
+
+  assign_string_constant(
+    state,
+    symex_assign,
+    to_ssa_expr(f_l1.arguments().at(0)),
+    new_char_array_length,
+    to_ssa_expr(f_l1.arguments().at(1)),
+    new_char_array);
+
+  return true;
+}
+
+bool goto_symext::constant_propagate_replace(
+  statet &state,
+  symex_assignt &symex_assign,
+  const function_application_exprt &f_l1)
+{
+  const auto &f_type = to_mathematical_function_type(f_l1.function().type());
+  const auto &length_type = f_type.domain().at(0);
+  const auto &char_type = to_pointer_type(f_type.domain().at(1)).subtype();
+
+  const refined_string_exprt &s = to_string_expr(f_l1.arguments().at(2));
+  const auto s_data_opt = try_evaluate_constant_string(state, s.content());
+
+  if(!s_data_opt)
+    return false;
+
+  auto &new_data = f_l1.arguments().at(3);
+  auto &old_data = f_l1.arguments().at(4);
+
+  // If it's a symbol we're using the single character override.
+  if (new_data.id() == ID_symbol)
+  {
+    const auto &oldChar = to_symbol_expr(new_data);
+    const auto &newChar = to_symbol_expr(old_data);
+
+    const auto new_char_pointer =
+      state.propagation.find(newChar.get_identifier());
+    const auto old_char_pointer =
+      state.propagation.find(oldChar.get_identifier());
+
+    if(!new_char_pointer || !old_char_pointer)
+    {
+      return {};
+    }
+
+    auto new_character = new_char_pointer->get().get(ID_value);
+    auto old_character = old_char_pointer->get().get(ID_value);
+
+    array_exprt string_data = s_data_opt->get();
+    auto &operands = string_data.operands();
+    for(auto &operand : operands)
+    {
+      if(operand.get(ID_value) == old_character)
+        operand.set(ID_value, new_character);
+    }
+
+    const constant_exprt new_char_array_length =
+      from_integer(string_data.operands().size(), length_type);
+
+    const array_typet new_char_array_type(char_type, new_char_array_length);
+    const array_exprt new_char_array(std::move(operands), new_char_array_type);
+
+    assign_string_constant(
+      state,
+      symex_assign,
+      to_ssa_expr(f_l1.arguments().at(0)),
+      new_char_array_length,
+      to_ssa_expr(f_l1.arguments().at(1)),
+      new_char_array);
+  }
+
+  return true;
+}bool goto_symext::constant_propagate_trim(
+  statet &state,
+  symex_assignt &symex_assign,
+  const function_application_exprt &f_l1)
+{
+  const auto &f_type = to_mathematical_function_type(f_l1.function().type());
+  const auto &length_type = f_type.domain().at(0);
+  const auto &char_type = to_pointer_type(f_type.domain().at(1)).subtype();
+
+  const refined_string_exprt &s = to_string_expr(f_l1.arguments().at(2));
+  const auto s_data_opt = try_evaluate_constant_string(state, s.content());
+
+  if(!s_data_opt)
+    return false;
+
+  array_exprt string_data = s_data_opt->get();
+
+  auto &operands = string_data.operands();
+  for(auto it = operands.begin();
+      it != operands.end();)
+  {
+    if(it->get_int(ID_value) == 20)
+      operands.erase(it);
```

Comment:
Don't think it makes a difference the way I'm using it here though?

Reply:
It might work though according to the documentation `it` becomes invalid when calling `erase(it)`. So I'd say it's safer to use the returned iterator.

Reply:
+1, it's just luck that it works here based on how `std::vector`'s iterators happen to work (i.e. being pointers to the data array, and the array not usually being reallocated on an erase). How about

```
auto expr_is_space = [](const exprt &e)
{
  const auto &constant_value = to_constant_expr(*it);
  return numeric_cast_v<int>(constant_value) <= ' ';
}

while(!operands.empty() && expr_is_space(operands.front()))
  operands.erase(operands.begin());
while(!operands.empty() && expr_is_space(operands.back()))
  operands.erase(operands.rbegin().base());
```

Reply:
Thinking about it the current implementation has n^2 worst case runtime due to the repeated shifting of the vector when erasing elements at the beginning. It might be better to use `find_if()` to get iterators to the first and last non-space characters, and then copy that segment to a new vector (instead of making a copy of the original vector and then modifying it).

Reply:
Sure, modified.